File: storage/db-core/src/main/java/com/coffee/gu/stamp/StampEntity.java
package com.coffee.gu.stamp;

import com.coffee.gu.Principal;
import com.coffee.gu.enums.PrincipalType;
import com.coffee.gu.enums.StampState;
import com.coffee.gu.BaseEntity;
import jakarta.persistence.*;

import java.time.LocalDateTime;

@Table(
        name = "stamp",
        indexes = {
                @Index(
                        name = "idx_stamp_user_state_expiry_created",
                        columnList = "principalKey, state, expiredAt, createdAt"
                )
        }
)
@Entity
public class StampEntity extends BaseEntity {
    private String orderKey;
    private String principalKey;
    @Enumerated(EnumType.STRING)
    private PrincipalType principalType;
    @Enumerated(EnumType.STRING)
    private StampState state;
    private LocalDateTime expiredAt;

    public StampEntity() {}

    private StampEntity(String orderKey, String principalKey, PrincipalType principalType, StampState state, LocalDateTime expiredAt) {
        this.orderKey = orderKey;
        this.principalKey = principalKey;
        this.principalType = principalType;
        this.state = state;
        this.expiredAt = expiredAt;
    }

    public static StampEntity from(Stamp stamp) {
        return new StampEntity(
                stamp.getOrderKey(),
                stamp.getPrincipal().getKey(),
                stamp.getPrincipal().getType(),
                stamp.getState(),
                stamp.getExpiredAt());
    }


    public LocalDateTime getExpiredAt() {
        return expiredAt;
    }

    public Stamp toModel() {
        return new Stamp(
                this.id,
                this.orderKey,
                new Principal(this.principalKey, this.principalType),
                this.state,
                this.getCreatedAt(),
                this.expiredAt);
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/stamp/StampEntity.kt
package com.coffee.gu.stamp

import com.coffee.gu.BaseEntity
import com.coffee.gu.Principal
import com.coffee.gu.enums.PrincipalType
import com.coffee.gu.enums.StampState
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Index
import jakarta.persistence.Table
import java.time.LocalDateTime

@Table(
    name = "stamp",
    indexes = [
        Index(
            name = "idx_stamp_user_state_expiry_created",
            columnList = "principalKey, state, expiredAt, createdAt",
        ),
    ],
)
@Entity
class StampEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0,
    val orderKey: String,
    val principalKey: String,
    @Enumerated(EnumType.STRING)
    val principalType: PrincipalType,
    @Enumerated(EnumType.STRING)
    var state: StampState,
    val expiredAt: LocalDateTime,
) : BaseEntity() {

    fun toModel(): Stamp {
        return Stamp(
            id = id,
            orderKey = orderKey,
            principal = Principal(principalKey, principalType),
            state = state,
            createdAt = createdAt,
            expiredAt = expiredAt,
        )
    }

    companion object {
        fun from(stamp: Stamp): StampEntity = StampEntity(
            id = stamp.id,
            orderKey = stamp.orderKey,
            principalKey = stamp.principal.key,
            principalType = stamp.principal.type,
            state = stamp.state,
            expiredAt = stamp.expiredAt,
        )
    }

}
